### PR TITLE
Beefy: small fixes

### PR DESCRIPTION
**Pull request by serban300**
Related to #2285

- save the state of the BEEFY gadget after processing a finality proof. We need this in order to avoid skipping blocks.
- avoid reprocessing the old state when not necessary

### CODE REVIEW

File: substrate/client/consensus/beefy/src/worker.rs [hunk 1]
```diff
@@ -478,6 +478,9 @@ where
 				}
 			}
 
+			crate::aux_schema::write_voter_state(&*self.backend, &self.persisted_state)
+				.map_err(|e| Error::Backend(e.to_string()))?;
+
```
Comment (acatangiu) [inline substrate/client/consensus/beefy/src/worker.rs:483]: for efficiency purposes I would move this to `fn init_session_at()`, because that's what we really care to persist - new detected sessions that are mandatory to vote on.

That way we avoid writing full persisted state to aux-db on every GRANDPA finality - usually this new state change leads to voting which already persists data.

TLDR: your code is the fully correct thing to do, but I'm suggesting we can relax a bit the data backup here and only do it for mandatory blocks finality (new sessions) instead of any block finality.

**Review by acatangiu — APPROVED** [review]
looks good! thanks for the fix!

File: substrate/client/consensus/beefy/src/worker.rs (hunk 1 above)
Comment (serban300) [reply substrate/client/consensus/beefy/src/worker.rs]: Makes sense. Added https://github.com/paritytech/polkadot-sdk/pull/2378/commits/51bac2f63e395a79704cc5b1c9e8f6f8a61bb414 . PTAL.

I added a flag for this and only saving when the flag is true. If we save inside `init_session_at()`, in theory we might save multiple times per notification. Even though I don't think it should happen in practice.

I'm also ok with saving inside `init_session_at()`  if you think it's better.

**Review by svyatonik — APPROVED** [review]

File: substrate/client/consensus/beefy/src/worker.rs (hunk 1 above)
Comment (acatangiu) [reply substrate/client/consensus/beefy/src/worker.rs]: > in theory we might save multiple times per notification

no, that should never happen, first block in each new session **has to be** GRANDPA finalized same as it has to be BEEFY finalized (mandatory block) - aka we will never have a single GRANDPA finality notification covering more than 1 session

(later on, we should really add some fn on `PersistedState` like `fn writer() -> Writer(&mut PersistedState)` that allows mutating `PersistedState` but also automatically persists it to AUX DB on `drop()`)

Comment (serban300) [reply substrate/client/consensus/beefy/src/worker.rs]: > (later on, we should really add some fn on `PersistedState` like `fn writer() -> Writer(&mut PersistedState)` that allows mutating `PersistedState` but also automatically persists it to AUX DB on `drop()`)

Agree. Would be really good to have such a mechanism.
